internal/state/graph/reference_grant_test.go

```diff
@@ -0,0 +1,201 @@
+						Group: "core",
+						Kind:  "Secret",
+						Name:  helpers.GetPointer(v1beta1.ObjectName(secretNsName.Name)),
+					},
+				},
+			},
+		}
+	}
+
+	createModifiedRefGrant := func(mod func(rg *v1beta1.ReferenceGrant)) *v1beta1.ReferenceGrant {
+		rg := getNormalRefGrant()
+		mod(rg)
+		return rg
+	}
+
+	tests := []struct {
+		refGrants map[types.NamespacedName]*v1beta1.ReferenceGrant
+		msg       string
+		allowed   bool
+	}{
+		{
+			msg: "allowed; specific ref grant exists",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "wrong-ns", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Namespace = "wrong-ns"
+				}),
+				{Namespace: "test", Name: "wrong-to-kind"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Spec.To[0].Kind = "WrongKind"
+					rg.Name = "wrong-to-kind"
+				}),
+				{Namespace: "test", Name: "rg"}: getNormalRefGrant(),
+			},
+			allowed: true,
+		},
+		{
+			msg: "allowed; all-namespace ref grant exists",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "test", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Spec.To[0].Name = nil
+				}),
+			},
+			allowed: true,
+		},
+		{
+			msg: "allowed; implicit to Group",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "test", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Spec.To[0].Group = ""
+				}),
+			},
+			allowed: true,
+		},
+		{
+			msg: "allowed; one matching 'to' ref",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "test", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Spec.To = []v1beta1.ReferenceGrantTo{
+						{
+							Group: "wrong.group",
+						},
+						{
+							Kind: "WrongKind",
+						},
+						{
+							Group: "core",
+							Kind:  "Secret",
+							Name:  helpers.GetPointer(v1beta1.ObjectName(secretNsName.Name)),
+						},
+					}
+				}),
+			},
+			allowed: true,
+		},
+		{
+			msg: "allowed; one matching 'from' ref",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "test", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Spec.From = []v1beta1.ReferenceGrantFrom{
+						{
+							Group: "wrong.group",
+						},
+						{
+							Kind: "WrongKind",
+						},
+						{
+							Group:     "gateway.networking.k8s.io",
+							Kind:      "Gateway",
+							Namespace: v1beta1.Namespace(gwNs),
+						},
+					}
+				}),
+			},
+			allowed: true,
+		},
+		{
+			msg: "not allowed; no ref group in secret namespace",
+			refGrants: map[types.NamespacedName]*v1beta1.ReferenceGrant{
+				{Namespace: "wrong-ns", Name: "rg"}: createModifiedRefGrant(func(rg *v1beta1.ReferenceGrant) {
+					rg.Namespace = "wrong-ns"
+				}),
+			},
+			allowed: false,
+		},
+		{
+			msg: "not allowed; no ref group with the right from Group",
```

Comment:
```suggestion
			msg: "not allowed; no ref group with the right 'from' Group",
```
Ditto changes for remaining tests.